docs/spec_draft.md

```diff
@@ -345,6 +346,57 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.bitcast_convert
+
+### Semantics
+
+Performs an element-wise bitcast operation from `operand` tensor and produces
+a `result` tensor where the bits are reinterpreted with the new element type.
+Bitcast is implemented as a low-level cast, so machines with different
+floating-point representations will give different results (e.g. endianesss).
+Conversions between different bitwidths are not element-wise and creates/deletes
+a dimension if new element type requires less/more bits, respectively.
+
+More formally, let `E` and `E'` be the `operand` and `result` element type,
+respectively:
+  * If `numBits(E) > numBits(E')`,
+    `result[i0, ..., iR-1, :] = operand[j0, ..., jR-1]` where
+    `R = rank(operand)`, and `dim(result, -1) = numBits(E)/numBits(E')`.
+  * If `numBits(E') > numBits(E)`,
+    `result[i0, ..., iR-2] = operand[j0, ..., jR-2, :]` where
+    `R = rank(operand)`, and `dim(operand, -1) = numBits(E')/numBits(E)`.
+
+### Inputs
+
+| Name      | Type                         |
+|-----------|------------------------------|
+| `operand` | tensor of any supported type |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the same shape except for the last
+    dimension.
+  * (C2) Cannot convert between real and complex type.
+
```

Comment:
max(bit_width(E),bit_width(E')) % min(bit_width(E),bit_width(E')) == 0
bit-widths should be perfectly divisible.

Reply:
If numBits(E') > numBits(E) then dim(operand, -1) = numBits(E') / numBits(E)

Reply:
> max(numBits(E), numBits(E')) % min(numBits(E), numBits(E')) == 0
> numBits should be perfectly divisible.

Since the current supported types are powers of 2, it will be perfectly divisible by default.

> If numBits(E') > numBits(E) then dim(operand, -1) = numBits(E') / numBits(E)

Thanks for the suggestion. I moved the entire shape inference to the constraints section.

Reply:
I think this belongs to the formal semantics, you can make 3 cases (add one case with equality)

Reply:
Let `E` and `E'` be the `operand` and `result` element type and `R = rank(operand)` respectively:
    * If `numBits(E) == numBits(E')`,
       `bitRepr(result[i0, ..., iR-1]) = bitRepr(operand[i0, ..., iR-1])`
        `shape(result) = shape(operand)`
    * If `numBits(E) > numBits(E')`,
      `bitRepr(result[i0, ..., iR-1, :]) = bitRepr(operand[i0, ..., iR-1])`
      `rank(result) = R+1` and `dim(result, -1) = numBits(E)/numBits(E')`
    * If `numBits(E') > numBits(E)`,
      `bitRepr(result[i0, ..., iR-2]) = bitRepr(operand[i0, ..., iR-2, :])`
      `rank(result) = R-1` and `dim(operand, -1) = numBits(E)/numBits(E')`

Reply:
If you put this in semantics you can remove C1

Reply:
Let me know what you think

Reply:
Note I have removed j0 ... and changed them to i0 ... because they are same

Reply:
I've updated the semantics and constraints. Please take a look and let me know any suggestions.